src/Generator/HtmlForPdfGenerator.php

```diff
@@ -45,8 +46,99 @@ public function generateHtmlForPdf(string $htmlDir, string $parseOnly/*, Progres
+                throw new \LogicException(sprintf('File "%s" does not exist', $filename));
+            }
+
+            $crawler     = new Crawler(file_get_contents($filename));
+            $fileContent = $crawler->filter('body')->html();
+
+            $uid = str_replace('/', '-', $meta->getFile());
+            $dir = dirname($meta->getFile());
+
+            // fix internal URLs
+            $fileContent = preg_replace_callback(
+                '/href="([^"]+?)"/',
+                function ($matches) use ($dir) {
+                    if ('http' === substr($matches[1], 0, 4) || '#' === substr($matches[1], 0, 1)) {
+                        return $matches[0];
+                    }
+
+                    $path = [];
+                    foreach (explode('/', $dir.'/'.str_replace(['.html', '#'], ['', '-'], $matches[1])) as $part) {
+                        if ('..' == $part) {
+                            array_pop($path);
+                        } else {
+                            $path[] = $part;
+                        }
+                    }
+
+                    $path = implode('-', $path);
+
+                    return sprintf('href="#%s"', $path);
+                },
+                $fileContent
+            );
+
+            // fix internal images
+            // $page = preg_replace('{src="(?:\.\./)+([^"]+?)"}', "src=\"$relativeImagesPath$1\"", $fileContent);
+
+            // fix # and id references to be unique
+            $fileContent = preg_replace_callback(
+                '/id="([^"]+)"/',
+                function ($matches) use ($uid) {
+                    return sprintf('id="%s-%s"', $uid, $matches[1]);
+                },
+                $fileContent
+            );
+
+            $content .= "\n";
+            $content .= sprintf('<div id="%s">%s</div>', $uid, $fileContent);
+        }
+
+        $content = sprintf(
+            '<html><head><title>%s</title></head><body>%s</body></html>',
+            $parseOnly,
+            $content
+        );
+
+        $content = $this->cleanupContent($content);
+
+        $filename = sprintf('%s/%s.html', $htmlDir, $parseOnly);
+        file_put_contents($filename, $content);
+        $fs->remove($basePath);
+    }
+
+    protected function cleanupContent($content)
+    {
+        // remove internal anchors
+        $content = preg_replace('#<a class="headerlink"([^>]+)>¶</a>#', '', $content);
+
+        // convert links to footnote
+        $content = preg_replace_callback(
+            '#<a href="(.*?)" class="reference external"(?:[^>]*)>(.*?)</a>#',
+            function ($matches) {
+                if (0 === strpos($matches[2], 'http')) {
+                    return sprintf('<em><a href="%s">%s</a></em>', $matches[2], $matches[2]);
+                }
+
+                return sprintf('<em>%s</em><span class="footnote"><code>%s</code></span>', $matches[2], $matches[1]);
+            },
+            $content
+        );
+
+        return $content;
```

Comment:
I would say that, instead of a functional test, this needs a good old-fashioned unit test :). Well, it will probably feel a bit like a functional test - "given this raw HTML, here is the cleaned up HTML" - but yes, 100% - this stuff needs a test.